_overviews/core/futures.md

```diff
@@ -338,47 +338,35 @@ a `List[String]`-- a list of recent textual posts:
 
 The `onComplete` method is general in the sense that it allows the
 client to handle the result of both failed and successful future
-computations. To handle only successful results, the `onSuccess`
-callback is used (which takes a partial function):
+computations. In the case where only successful results need to be
+handled, the `foreach` callback can be used:
 
     val f: Future[List[String]] = Future {
       session.getRecentPosts
     }
 
-    f onSuccess {
-      case posts => for (post <- posts) println(post)
+    f foreach { posts =>
+      for (post <- posts) println(post)
     }
 
-To handle failed results, the `onFailure` callback is used:
+To handle failed results, the `foreach` callback can be used on the
+`Future[Throwable]` obtained via the `failed` projection (which is
+explained [below](#projections)):
 
     val f: Future[List[String]] = Future {
       session.getRecentPosts
     }
 
-    f onFailure {
-      case t => println("An error has occured: " + t.getMessage)
+    f.failed foreach { t =>
+      println("An error has occured: " + t.getMessage)
     }
 
-    f onSuccess {
-      case posts => for (post <- posts) println(post)
+    f foreach { posts =>
+      for (post <- posts) println(post)
     }
 
-The `onFailure` callback is only executed if the future fails, that
-is, if it contains an exception.
-
-Since partial functions have the `isDefinedAt` method, the
-`onFailure` method only triggers the callback if it is defined for a
-particular `Throwable`. In the following example the registered `onFailure`
-callback is never triggered:
-
-    val f = Future {
-      2 / 0
-    }
-
-    f onFailure {
-      case npe: NullPointerException =>
-        println("I'd be amazed if this printed out.")
-    }
+The `failed.foreach` callback is only executed if the future fails, that
```

Comment:
That's technically true, but doesn't really explain what's going on. It'd be more accurate to say that `future.failed` converts the `Future[Failed[Throwable]]` to a `Future[Success[Throwable]]` to allow `foreach` to operate on the `Throwable`, and will produce a `Future[Failure[NoSuchElementException]]` if the original future succeeded, essentially swapping the position of the values.

I think probably the wording can just be a little simpler though, especially as there's a link to a projections anchor further down the page.

"To handle failed results, you can first use the `failed` projection to convert the `Failure[Throwable]` to a `Success[Throwable]` and then use `foreach` on the newly-successful `Future` instead."

I do note that the projections section clarifies that the `failed` projection is to enable handling exceptions with for comprehensions, though, so I wonder if there's actually an advantage to `future.failed.foreach` over simply

```scala
future onComplete {
  case Success(_) =>
  case Failure(t) => ... 
```
given that the former requires creating a new `Future`.

Reply:
I agree that `failed.foreach` doesn't really add very much (indeed if this was particularly useful then I expect `onFailure` wouldn't be deprecated).

So how about we just remove this failure-scenario discussion/example, replace it with your "to handle failed results" paragraph and a link to the 'Projections' section where it can be properly explained?

Reply:
That makes sense to me.

Reply:
Okay so I've went a bit further to try and make things immediately obvious if the link is followed. See what you think..